app/scripts/controllers/transactions/index.js

```diff
@@ -1646,136 +1395,519 @@ export default class TransactionController extends EventEmitter {
+    if (actionId) {
+      txMeta.actionId = actionId;
+    }
+
+    if (origin === ORIGIN_METAMASK) {
+      // Assert the from address is the selected address
+      if (normalizedTxParams.from !== this.getSelectedAddress()) {
+        throw ethErrors.rpc.internal({
+          message: `Internally initiated transaction is using invalid account.`,
+          data: {
+            origin,
+            fromAddress: normalizedTxParams.from,
+            selectedAddress: this.getSelectedAddress(),
+          },
+        });
+      }
+    } else {
+      // Assert that the origin has permissions to initiate transactions from
+      // the specified address
+      const permittedAddresses = await this.getPermittedAccounts(origin);
+      if (!permittedAddresses.includes(normalizedTxParams.from)) {
+        throw ethErrors.provider.unauthorized({ data: { origin } });
+      }
+    }
+
+    const { type } = await determineTransactionType(
+      normalizedTxParams,
+      this.query,
+    );
+    txMeta.type = transactionType || type;
+
+    // ensure value
+    txMeta.txParams.value = txMeta.txParams.value
+      ? addHexPrefix(txMeta.txParams.value)
+      : '0x0';
+
+    if (txMethodType && this.securityProviderRequest) {
+      const securityProviderResponse = await this.securityProviderRequest(
+        txMeta,
+        txMethodType,
+      );
+
+      txMeta.securityProviderResponse = securityProviderResponse;
+    }
+
+    this.addTransaction(txMeta);
+
+    txMeta = await this.addTransactionGasDefaults(txMeta);
+
+    if (
+      [TransactionType.swap, TransactionType.swapApproval].includes(
+        transactionType,
+      )
+    ) {
+      txMeta = await this._createSwapsTransaction(
+        options?.swaps,
+        transactionType,
+        txMeta,
+      );
+    }
+
+    return { txMeta, isExisting: false };
+  }
+
+  async _createSwapsTransaction(swapOptions, transactionType, txMeta) {
+    // The simulationFails property is added if the estimateGas call fails. In cases
+    // when no swaps approval tx is required, this indicates that the swap will likely
+    // fail. There was an earlier estimateGas call made by the swaps controller,
+    // but it is possible that external conditions have change since then, and
+    // a previously succeeding estimate gas call could now fail. By checking for
+    // the `simulationFails` property here, we can reduce the number of swap
+    // transactions that get published to the blockchain only to fail and thereby
+    // waste the user's funds on gas.
+    if (
+      transactionType === TransactionType.swap &&
+      swapOptions?.hasApproveTx === false &&
+      txMeta.simulationFails
+    ) {
+      await this._cancelTransaction(txMeta.id);
+      throw new Error('Simulation failed');
+    }
+
+    const swapsMeta = swapOptions?.meta;
+
+    if (!swapsMeta) {
+      return txMeta;
+    }
+
+    if (transactionType === TransactionType.swapApproval) {
+      this.emit('newSwapApproval', txMeta);
```

Comment:
What's the rationale for emitting an event here instead of just calling the swaps controller method here?

Reply:
Oh, maybe because the transactions controller has to be instantiated before the swaps controller?

Reply:
Exactly, a circular dependency. We would need the `SwapsController` to exist first to provide the callback in the constructor, but the `SwapsController` requires the `TransactionController` for callbacks in its constructor.

An event enables us to create the listener after both are already constructed.